### DISCUSSION TITLE
What language proposals would benefit from CLR changes?

### DISCUSSION THREAD

**Discussion by gafter** [dotnet/csharplang#317] [General]
If we were to plan for one language feature that requires a revision of the CLR, then we might as well do as many of them at the same time as make sense. What changes would we consider that would benefit from CLR support? This is a pared-down list for us to select from.

The most likely features we might do soon:
1. Default interface methods/traits (see https://github.com/dotnet/csharplang/issues/52)
2. Array slicing (see https://github.com/dotnet/csharplang/issues/185)

Other features that would benefit from CLR changes
1. CLR unification of types across assemblies (is there an issue?).
1. Make void a first-class type with one value; see https://github.com/dotnet/roslyn/issues/234 (Suggested by @ashmind) (Given how long the CLR has been around, it is probably too late for APIs to benefit, as those that would benefit most have already been written)
1. Covariance and contravariance for classes (see https://github.com/dotnet/roslyn/issues/171)
1. Allow |, &, and ~ operators on a type parameter with the enum constraint (see https://github.com/dotnet/roslyn/issues/262)
1. Union and intersection types (https://github.com/dotnet/csharplang/issues/344, though likely not that syntax).
1. Support generic indexers (see https://github.com/dotnet/roslyn/issues/523)
1. Higher-kinded polymorphism (see https://github.com/dotnet/csharplang/issues/339 and https://github.com/dotnet/coreclr/issues/6524)
1. Permit methods to be declared in enums (https://github.com/dotnet/csharplang/issues/297)

**Comment by louthy** [comment]
> Higher-kinded polymorphism (see dotnet/roslyn#2212)

Please!

**Comment by tannergooding** [comment]
@gafter, what about 'runtime constants for blittable structs'? That is, being able to declare a constant value for anything that has a constant bit layout (such as `Vector2`, `Vector3`, and `Vector4`).

**Comment by miloush** [comment]
Not sure if covered by #164 but I guess extending a type to implement an interface might benefit from CLR changes.

**Comment by SamPruden** [comment]
Well for starters I'll push #169 and #252, both suggestions around the inclusion of some form of thistype feature. Between them those issues have shown a significant interest in the feature, and I believe that would add a lot of power and niceness to the language.

Along with shapes and a few other half formed ideas I should hurry up and post, I'd really like to see this as the Generics 2.0 update.

**Comment by louthy** [comment]
@TheOtherSamP Not that I know of, I was just quoting @gafter.  

> I'd really like to see this as the Generics 2.0 update.

Indeed.  I think the value of higher-kinds is so large that it does deserve one of those _once in a decade_ major changes to the CLR (the last change _of this size_ was to enable generics 1.0, I believe?  although maybe the DLR was as significant).  

There are so many features that have been proposed that are skirting around this issue (concepts, shapes, default interface methods, etc.), it seems crazy to not do this properly.  It's not even like it's a poorly understood concept.  The constant mantra of _"we can't touch the CLR"_ is understandable (to an extent), but has been frustrating when you see the language hit the limits of its apparently immutable runtime.  So it's nice to see this question pop up.

I'd love to see a more powerful constraints system also.

**Comment by SamPruden** [comment]
> The constant mantra of "we can't touch the CLR" is understandable (to an extent), but has been frustrating when you see the language hit the limits of its apparently immutable runtime. So it's nice to see this question pop up.

It is, though I'm not sure how much it really means given that it's basically a reposting of an old issue from 2015. Whether this is an indication that anything may happen in the near future isn't clear.

While I'm in the habit of shamelessly promoting my own issues on here, I'd like to point out #255 where I asked about the status of the CLR. That received quite a bit of support, it's currently sitting at 15 thumbs, but this issue is the closest thing we've got to a response so far.

**Comment by BreyerW** [comment]
I would like to see more constrains like:

-`unmanaged`
-`blittable`
-flexible constructor constraint like `new (int, string)` (would be great if applying multiple different constructor constraint was possible)
- maybe constraint based on attributes?

I think every bullet have proposal but im writing it from phone so sorry for no exact information, will update late.

Could it be possible to make static members as part of interface contract? Since you are planning on extending interface with default members and enable static methods with body it woould be unintuitive to be able write defalt static methods and instance members and make contract on instance members but no static members.

For that there is no proposal, i can make one or post it in default member proposal

Additionally i would like to mention my proposal that can introduce some form of intersection type based on already existing construct: dotnet/roslyn#7644 i think it may go well with higher-kind polymorphism?

**Comment by ufcpp** [comment]
What about:

- [Generic Attributes](https://github.com/dotnet/csharplang/issues/124)
- [Covariant Return Types](https://github.com/dotnet/csharplang/blob/master/proposals/covariant-returns.md) looks like not necessarily require CLR change, but could this be better if CLR changes is allowed?

**Comment by ygc369** [comment]
@gafter 
@tannergooding 
I have some proposals that only need CLR changes and do not need new language feature. Thus we could get performance improvement without any code change.

- https://github.com/dotnet/coreclr/issues/555: optimization for LOH GC
- https://github.com/dotnet/coreclr/issues/757: optimization for Math.DivRem
- https://github.com/dotnet/coreclr/issues/1784: stack allocation(Escape Analysis)
- https://github.com/dotnet/coreclr/issues/1856: optimization for finalizer

**Comment by whoisj** [comment]
Support for `struct`. Yes, I know that `ValueTypes` are supported, but I really want RAII like handling of stack allocated objects. I want it so bad. Borrow operators can be implemented at the language/compiler level but proper stack allocated objects (with c'tor/d'tor) support is something we need the CLR to support.

I will absolutely relish the day when I can write an entire application and not ever worry about GC behavior because nearly everything I care about is stack allocated and/or shared via borrow semantics. :grin:

**Comment by DavidArno** [comment]
For me, the four most important would be:

1. [Overridable default operator](https://github.com/dotnet/csharplang/issues/167) to better support non-nullable reference types and structs with non-zero default values.
2. Direct CLR support for unions/ADTs (would make implementing them in C# easier and would likely benfit F# too).
3. `void` as a returnable type when required, to remove the need for a `unit` type (that could just be an alias for `void`) and the need for handling actions and functions as separate delegates.
4. Faster delegates, ie just treating them as direct references to static functions (including non-capturing lambdas), though [static delegates](https://github.com/dotnet/csharplang/issues/302) might already pretty much cover that without the need for CLR changes.

**Comment by alrz** [comment]
- Allow `T?` for an unconstrained generic type.
- Implement interfaces for existing classes (virtual extension methods)
- CLR support for type aliases https://github.com/dotnet/csharplang/issues/259 to make them public.
- Lowerbound type constraints

>Covariance and contravariance for classes

- I'd love to have covariant tuples out-of-the-box.
- Covariant `out` parameters, anyone?

Also, with traits, we might need a new kind of constraints, for example,
```cs
void F<T>() where X<T> : Trait
```
Here, we constrained `T` such that there exists an implementation of `Trait` for `X<T>`.

**Comment by MikePopoloski** [comment]
> Support for struct. Yes, I know that ValueTypes are supported, but I really want RAII like handling of stack allocated objects. I want it so bad. Borrow operators can be implemented at the language/compiler level but proper stack allocated objects (with c'tor/d'tor) support is something we need the CLR to support.
> 
> I will absolutely relish the day when I can write an entire application and not ever worry about GC behavior because nearly everything I care about is stack allocated and/or shared via borrow semantics. 😁

As someone who has been using C# for game development scenarios for a *long* time, this is the thing that would be most beneficial to me. Fancy language features are nice, but this would be a game changer.

I should also acknowledge that there's definitely been progress made already; Vector types, the Unsafe class, and ref returns were all very welcome.

**Comment by ufcpp** [comment]
@whoisj 
RAII doesn't require CLR changes.
https://github.com/dotnet/roslyn/issues/161

**Comment by john-cd** [comment]
I would second @whoisj  regarding borrowing semantics. 

Did the CLR team have a look at the Rust language and its concept of ownership, borrowing and life times?

https://www.toptal.com/software/eliminating-garbage-collector

**Comment by john-cd** [comment]
What about CLR support for code contracts - requires / ensures ... - via static checking?

Useful for C# and VB

**Comment by john-cd** [comment]
What about custom operators?

Swift example:

extension Vector2D {
    static prefix func +++ (vector: inout Vector2D) -> Vector2D {
        vector += vector
        return vector
    }
}
 
var toBeDoubled = Vector2D(x: 1.0, y: 4.0)
let afterDoubling = +++toBeDoubled
// toBeDoubled now has values of (2.0, 8.0)
// afterDoubling also has values of (2.0, 8.0)

**Comment by AlgorithmsAreCool** [comment]
@john-cd Custom operators don't need CLR support. F# has had them for years.

**Comment by stepanbenes** [comment]
`never` type and support for methods that never return requires CLR support. see: dotnet/roslyn#1226

**Comment by john-cd** [comment]
Could you consider adding low-level features to improve concurrent programming? 

I am thinking of
- better async / multithreaded support (e.g. async is not allowed in lock block)
- truly immutable variables to avoid e.g.  readonly C v = new C();  v.x = 42; if C has not been properly designed
- higher-level primitives in the Interlocked class: the lock-free optimistic concurrency check pattern with compareexchange is difficult 
- transactional memory:
   https://en.wikipedia.org/wiki/Transactional_memory
   http://joeduffyblog.com/2010/01/03/a-brief-retrospective-on-transactional-memory/

**Comment by alrz** [comment]
I'd like to add struct ADTs to the list. while class ADTs can be represented as a class hierarchy, struct ADTs should be represented as unions. Currently `FieldOffset` does not support generics which is a blocker for said feature (if we intend to translate such types to structs with an explicit layout).

**Comment by tannergooding** [comment]
@alrz, I think the primary issue with `FieldOffset` and generics is that their is no current way to constrain it to only allow types with a fixed bit layout (that is, constrain it to be 'blittable'). If we had that support, then I think FieldOffset would automatically work for generics constrained as such.

**Comment by alrz** [comment]
@tannergooding I had intended that to be unconstrained though. A possible optimization on clr part would be to not allocate a flag when the union consists of an empty case and a reference type. That's how Rust represents `Option<T>` union. An instance of `Option<&T>` doesn't need a flag since all zeros effectively means the empty case, i.e. null.

**Comment by tannergooding** [comment]
@alrz, but what does it mean for two heap-allocated objects to exist in the same location?

For example, if I declared:
```
[StructLayout(LayoutKind.Explicit)]
public struct SomeStruct<T, U>
{
    [FieldOffset(0)]
    public T _t;

    [FieldOffset(0)]
    public U _u;
}
```

U and T, which are separate heap allocated objects, would share the same location in the struct (which can only happen if both values are always null).

**Comment by alrz** [comment]
@tannergooding Your example is not fit for ["null-pointer optimization"](https://github.com/rust-unofficial/too-many-lists/blob/master/src/first-layout.md) that I mentioned which makes it unsized at the time of declaration. It depends on the "instants" of the said types, if both `T` and `U` are reference types they shall not exist in the same location. That is probably not how clr works as of today which is why I mentioned it here.

**Comment by tannergooding** [comment]
@alrz, it isn't just if both T and U, its if either T or U are heap-allocated object, then neither can exist in the same location. For example, `SomeStruct<object, IntPtr>` is illegal. Even though both exist in the same location, both are the same size, and they will never point to separate objects.

That is what the `blittable` constraint does. It basically says, I contain no heap-allocated objects and have a constant bit layout. Therefore, I can exist in the same memory location as any other blittable type, I require no marshalling, etc, etc.

**Comment by RandyBuchholz** [comment]
Rethink `Enum` in general. I've followed several threads around enums and it seems that a lot of them get stalled out because of CLR limitations.

**Comment by tannergooding** [comment]
@RandyBuchholz, I think the current biggest limiter on Enum, is that we don't support more than a handful of literal (that is runtime constant) types. If we supported being able to declare literals for any blittable type, then Enum could be naturally extended to support any type for which you could declare a constant.

**Comment by RandyBuchholz** [comment]
@tannergooding Would that include being able to create nested enums?

**Comment by tannergooding** [comment]
@RandyBuchholz, No. I think being able to do things like: nest classes, or declare methods in an enum would be separate (their is already a championed issue on the latter https://github.com/dotnet/csharplang/issues/297).

**Comment by tannergooding** [comment]
@alrz, after going over it a bit more (and this may be just because I'm not familiar with Rust and/or not understanding). It seems like, (in the example given) this is used to optimize away a flag that describes what the type is.

For example, in Rust an object cannot be null (as I understand it). So, you declare `enum Link { Empty, Box<Node> }`, which essentially says "I can be either type Empty or type Box\<Node\>".
Normally, this would be internally implemented as a data structure, something like `struct Link { uint type; uint data; }`. Where Empty would be `{ type = 0; data = nothing }` and `Box<Node>` would be `{ type = 1; data = link }`. Rust will then perform null-pointer optimization to say, `{ data = 0 }` is Empty and `{ data = anything-else }` is `Box<Node>` (since link can never be 0).

However, in C#, Node can be null. So you just need to declare `class Node<T> { T data; Node next; }`. and it takes up the same size, without having to carry around a bit to declare the type.

Additionally, the type information is carried around by the CLR and never actually directly exposed in the structures passed around. It is hidden away by the runtime and inlined where appropriate.

**Comment by whoisj** [comment]
@tannergooding I fail to see how `Empty` isn't just another `null`. A dereference of `Empty` most likely ends in an unhappy state. Rust avoids `null` by making heap allocation frustrating (thus encouraging stack allocations). C# is quite the opposite - each has their place. Forcing C# to non-null isn't necessarily a win.

I will say that something like `class Node<T> { T data; Node next; }` is a neat trick, and would be rather useful. Kind of like a safely type `void*`. 🤤

**Comment by alrz** [comment]
@tannergooding 


>However, in C#, Node can be null.

Not if it's a struct or a non-nullable reference type. The end game is to allow `T?` for an unconstrained `T`. With current implementation we still require `struct` or `class` constraints which is inconvenient. In fact I think `struct` and `class` constraints no longer make sense, because we can have a non-nullable `class` but it cannot satisfy a `struct` constraint which effectively means "non-nullable".

Imagine if we want to remove `struct` constraint for `Nullable<T>` and redeclare it as a struct ADT,
```cs
enum struct Nullable<T> {
  Some(T),
  None;
  // ... existing Nullable<T> members for backward-compat
}
```
Here, `T` in `Some(T)` may not be null because it's not defined as such. We also cannot have `T??` for the same reason. While  `RefType` and `RefType?` are the same when we translate them to IL (perhaps just with a `NullableAttribute` in case of a `RefType?`), in order to define an unconstrained `T?` we should remove `struct` constraint to be able to hold both reference and value types as nullables. This brings us to null-pointer optimization, when we pass a `RefType?` we want to eliminate the space needed for the tag since it's just additional overhead, but it can't be done on the compiler side.

PS: There is a flaw in my example because of constraint namings. Unconstrained generics may accept nullables, but for this particular case `T` may not, short of being a struct or class, thus it requires special treatment in the language, but it could be generalized on clr part.

**Comment by svick** [comment]
@john-cd 

> better async / multithreaded support (e.g. async is not allowed in lock block)

Why would that need CLR support? `await` is not allowed in a `lock` by the C# compiler, because it's almost always a bug (`lock` is based on threads, `await` often switches threads). If you want `async`-compatible synchronization primitives, then those already exist: there's `SemaphoreSlim.WaitAsync()` and if you want more, you can use [primitives from Stephen Cleary's AsyncEx](https://github.com/StephenCleary/AsyncEx/wiki#asynchronous-coordination-primitives). If you want special syntax for some of those primitives, then that's a matter of modifying the C# compiler and possibly adding them to CoreFX/.Net Standard, no CLR changes are necessary.

**Comment by sharwell** [comment]
While the consensus is covariant return types (dotnet/roslyn#357, dotnet/roslyn#311) wouldn't require CLR changes, it's possible they could benefit from a CLR change. It addresses a few items related to inherited attributes and could also reduce the size of the compiled assembly.

**Comment by john-cd** [comment]
@svick  I should have been clearer:  yes, ``await`` inside a ``lock`` is 99% of the time a bug. 

My points are the following:
- Locking primitives (``Monitor``, ``Interlocked``, newer <>Slim classes, ``MemoryBarrier``, ``volatile`` keyword...) are often thin veneer on top of OS primitives or CPU features.
- They offer no uniform API.
- They are difficult to use and even harder to debug.
- The TPL approach simplifies things, yet concurrent programming (``await``) and multithreading features (``Task.Wait()``) are jumbled together. ``SemaphoreSlim.WaitAsync()`` is another example.
- There is a need for a new, clean, low-level features implemented in the CLR, such as deep-immutability (as opposed to ``readonly`` shallow immutability), perhaps ``atomic { ... }`` memory transactions, 
- There is a need for built-in asynchronous equivalents to ``Monitor``, ``MemoryBarrier``... I like Stephen's work, but these primitives belong to the core Fx.  
- There is a need for C# language support for concurrency control keywords (similar to ``lock``) wrapping these primitives.

**Comment by john-cd** [comment]
Another suggestion:  improve SIMD support and allow GPGPU access. I would love to have a ``Scheduler`` / ``ThreadPool`` that allows me to transparently launch compute-intensive tasks on the GPU via ``Task.Run`` without the need to muck with P/Invoke and custom C /C++ code for CUDA or OpenCL... That is a hard job, which requires an IL-to-GPU compiler and memory transfer management but that has been done: http://www.quantalea.com/ 

Vectorized operations and/or GPU support and/or good integration to C / C++ engines that can do that is pretty essential for machine learning and esp. deep learning. This is why C# has quasi-zero presence in that space (ML is primarily coded in Python, Java, R, scala, even Julia...).

**Comment by tannergooding** [comment]
@john-cd, Neither of those requires a change to IL instruction set or CLR metadata, strictly speaking.

SIMD support already exists today, and is constantly being improved. GPGPU could also be provided in a similar manner.

I believe this issue is trying to list out any language features which are not implementable because of strict limitations in the IL instruction set or of what is allowed to be defined in CLR metadata.

**Comment by tannergooding** [comment]
> Not if it's a struct or a non-nullable reference type. The end game is to allow T? for an unconstrained T. With current implementation we still require struct or class constraints which is inconvenient. In fact I think struct and class constraints no longer make sense, because we can have a non-nullable class but it cannot satisfy a struct constraint which effectively means "non-nullable".

@alrz, I think the difference here is that, for `non-nullable` class, you still have a pointer under the hood (classes are always reference types, regardless of whether they are nullable or not). However, for structs (nullable or not), you don't have a pointer (structs are always value types).

There is some overlap for ref structs as compared to classes, since they are always a reference type, but I don't think that distinction is important here. The JIT should be able to optimize it away (as it already does with various generics in a lot of cases) without any explicit coding on users side.

**Comment by svick** [comment]
@john-cd 

> They offer no uniform API.

I'm not sure uniform API is possible, considering many of the primitives do very different things. Are you talking about something similar to `WaitHandle`, which is the common base class of `Mutex`, `Semaphore` and `Manual`-/`AutoResetEvent`? That sounds like something that could be considered when designing a new set of primitives or added to existing primitives.

> The TPL approach simplifies things, yet concurrent programming (`await`) and multithreading features (`Task.Wait()`) are jumbled together.

I doubt that's going to change anytime soon. `Task` is not perfect, but switching to something different would require significant improvements for it to be worth it, and I think just "unjumbling" is not enough.

(On the other hand, such type could be awaitable and tasklike, so adding it shouldn't require changes to the C# compiler.)

> There is a need for built-in asynchronous equivalents to `Monitor`, `MemoryBarrier`... I like Stephen's work, but these primitives belong to the core Fx.

Then you might want to consider creating a proposal in the CoreFx repo, following their [API review process](https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/api-review-process.md). If the implementation is going to require native code or coordination with the CLR, that's fine, as long as it looks just like a set of new classes from the point of view of the C# compiler.

**Comment by john-cd** [comment]
@tannergooding  

Noted. 

As far as I am aware, there are no SIMD instructions in the CIL language? 
Is this something that is considered?

**Comment by tannergooding** [comment]
@john-cd: https://github.com/dotnet/corefx/tree/master/src/System.Numerics.Vectors/src. All of these APIs compile down to raw SIMD instructions on supported platforms (x86, x64, ARM, etc..), but fallback to software implementation on platforms where hardware support isn't available.

There are still some perf issues, and not every intrinsic exposed to C/C++ is currently available, but it is more than sufficient for the majority of use-cases today.

There are multiple bugs tracking adding the additional intrinsics and improving performance to better match what you can implement using raw intrinsics in C/C++.

**Comment by john-cd** [comment]
@tannergooding   I am asking about introducing SIMD-like instructions to CIL, not the use of x86 SIMD (SSE, AVX...) instructions by the RyuJIT compiler. 

The C# compiler knows about the high-level loop structure, and can figure out the presence of vectors / matrices / tensors more easily than the JIT compiler can.   

At a minimum, the C# compiler could insert hints for the JIT compiler (perhaps via a CIL instruction similar to ``volatile.`` ) to suggest that a code section be x86 SIMD vectorized (if possible).  

A similar flag could be used to indicate preference for GPGPU offloading.

**Comment by tannergooding** [comment]
@john-cd, yes, leaving hints could help where the JIT currently misses things. 

However, a number of things, like hints, could easily be implemented (without requiring changes to the CLI) using either `Attributes` or via a special `CompilerHints` class that exposes a bunch of  methods for the various hint types. For example:
```
public static class CompilerHints
{
    [MethodImplAttribute(MethodImplOptions.InternalCall)]
    public static extern void Vectorize();

    [MethodImplAttribute(MethodImplOptions.InternalCall)]
    public static extern void OffloadToGPU();
}
```

Additionally, if performance is really that key, using an API which explicitly exposes the functionality (such as `System.Numerics.Vector4` and `System.Numerics.Matrix4x4`) is always better, as it guarantees it will be used.

**Comment by john-cd** [comment]
Point taken.  

As a side note, the GetHashCode implementation of some of the ``Matrix...`` types I looked at in the source code of System.Numerics.Vector is symmetric with respect to matrix coefficients. 

```csharp
public override int GetHashCode()
        {
            return unchecked(M11.GetHashCode() + M12.GetHashCode() +
                             M21.GetHashCode() + M22.GetHashCode() +
                             M31.GetHashCode() + M32.GetHashCode());
        }
```

That means that inserting a matrix, its transpose and some rotations of that matrix in a hashtable will result in collisions and a degradation from O(1) to O(n log n) or O(n) performance, depending on what collision resolution mechanism is used.

**Comment by lachbaer** [comment]
@gafter 
Please also see [(Proposal) `readonly` methods and properties for initialization [CLR change] #348](https://github.com/dotnet/csharplang/issues/348)

**Comment by orthoxerox** [comment]
@john-cd you should raise this issue in coreFX repo, since it is a library issue, not something that deeply affects the execution engine.

**Comment by whoisj** [comment]
Given than .Net Core native compiles, how far away are we from being able to `extern "c"` unsafe static functions, and other goodies?

What CLR changes, if any, would need to be made to make this possible?

**Comment by fubar-coder** [comment]
This would also benefit from a CLR change: `NotNullable<T>` struct that doesn't permit `default(T)` and throws an exception if accessed using `ldelem` when the element wasn't initialized.

**Comment by whoisj** [comment]
Gah! @terrajobst set me straight on this: "given than .Net Core native compiles". Seems that's not true. ☹️ Very 😢 🐼 - guess I need to go look at Rust lang again... 

💡 What would it take to give `struct` a deconstructor ❓

**Comment by masonwheeler** [comment]
[Metaclasses](https://github.com/dotnet/coreclr/issues/990) would be incredibly helpful.  They would eliminate the need for `Activator.CreateInstance` almost completely, greatly simplify Factory Pattern implementations, and allow a lot of what's currently done with Reflection to be done statically, with the corresponding benefits in performance and compile-time verifiability.

**Comment by dcnetfan** [comment]
'**this**' generic type constraint https://github.com/dotnet/csharplang/issues/169

**Comment by 4creators** [comment]
Another proposal related to generics #749 `constexpr` parametrized types.

**Comment by masonwheeler** [comment]
Not a proposal from elsewhere, just an idea I've been thinking about.

When you look at generator methods and `async/await`, there's a huge amount of bookkeeping going on that the compiler has to do in order to turn the method into a state machine that can be re-entered and resume at whatever arbitrary point you left it at, to preserve the polite fiction that you're using a coroutine that's pausing and resuming.

In Python, none of this happens, because stack frames are Python objects, and the `yield` operation saves the current stack frame object in memory, along with a value indicating the current instruction pointer.  When you come back to the next iteration of the generator, it just loads up that stack frame, jumps to the instruction pointer, and keeps going.

It shouldn't be that difficult to greatly simplify our coroutine operations by doing the same thing, but it would require three changes in the CLR:

1) a `stackframe` data type that holds the state of the relevant CPU registers, the instruction pointer, and an arbitrary number of bytes of stack space.  (Exact implementation would vary by platform, of course.)
2) a `SAVESTACK` IL instruction that saves the current stack frame to a field of type `stackframe`
3) a `LOADSTACK` IL instruction that loads the current stack frame state from a field of type `stackframe`.

If these were available, coroutines could be made much more performant by removing the state machine crud that clogs them up.

**Comment by ufcpp** [comment]
I think structs could be covariant with certain manner:

```cs
using System;
using System.Collections.Generic;
using System.Runtime.CompilerServices;

class Base { }
class Derived : Base { }

// Propssal: A struct all of whose fields are covariant could be covariant
struct CovariantStruct<T>
{
    public T Value;
    public IEnumerable<T> Items;
    public (bool isValid, T value) Optional;

    // Function members don't matter.
    // `T` can be used on parameters.
    public void Set(T value) => Value = value;
}

class Program
{
    // Force to bitwise-copy with the Unsafe class.
    static CovariantStruct<Base> BitwiseCopy(CovariantStruct<Derived> d) => Unsafe.As<CovariantStruct<Derived>, CovariantStruct<Base>>(ref d);

    static void Main(string[] args)
    {
        CovariantStruct<Derived> d = new CovariantStruct<Derived>
        {
            Value = new Derived(),
            Items = new List<Derived> { new Derived(), new Derived() },
            Optional = (true, new Derived()),
        };

        // So far, conversion from CovariantStruct<Derived> to CovariantStruct<Base> requires unsafe context.
        // However, this is safe as shown in the folloing.
        CovariantStruct<Base> b = BitwiseCopy(d);

        // All field accesses are safe (covariant).
        Console.WriteLine(b.Value);
        foreach (var item in b.Items) Console.WriteLine(item);
        Console.WriteLine(b.Optional.value);

        // This rewrite is safe because `b` is a copied value.
        b.Set(new Base());
        b.Items = new Derived[0];
        b.Optional = (false, new Base());

        // The original value `d` is not changed.
        Console.WriteLine(d.Value);
        foreach (var item in d.Items) Console.WriteLine(item);
        Console.WriteLine(d.Optional.value);
    }
}

// Open Question:
// How should "covariant struct" be defined

// 1. explicit or implicit?

// Version 1: This can be covariant.
struct X<T>
{
    public T Value { get; set; }
}
interface I<out T>
{
    // OK
    X<T> GetX();
}
// ↓
// Version 2: This is no longer covariant.
struct X<T>
{
    private StrongBox<T> _box;
    public T Value { get => _box.Value; set => _box.Value = value; }
}
interface I<out T>
{
    // NG. breaking change
    X<T> GetX();
}

// Should the language introduce some modifier? such as `struct X<out T>` (similar to blittable types)

// 2. Syntax

// 2-1. `out` modifier on type parameters?
struct X<out T>
{
    // This can be covariant because its field is covariant
    public T _value;

    // even though the type parameter `T` is used as parameters and set-accessors.
    public T Value { get => _value; set => _value = value; }
    public void SetValue(T value) => _value = value;
}
interface IVariant<out T, in U>
{
    CovariantStruct<T> GetCovariantValue();
    void SetContravariantValue(CovariantStruct<U> value);
}
// pros: Consistent with the modifier for interface/delegate.
// pros: No new keyword is needed.
// cons: It's not actually "out". Both "in" and "out" are allowed on function members.

// 2-2. `out` modifier on structs?
out struct X<T1, T2>
{
    public T1 Item1;
    public T2 Item2;
}

out struct Y<T1, T2>
{
    // covariant with T1
    public T1 Item1;

    // but invariant with T2
    public StrongBox<T2> Item2; // should be error. how?
}

// 2-3. other keywords?
struct X<covariant T1, T2>
{
    public T1 Item1;
    public StrongBox<T2> Item2;
}
struct X<variant T1, T2>
{
    public T1 Item1;
    public StrongBox<T2> Item2;
}
```

**Comment by jnm2** [comment]
Seems too magic and opaque. Things that are semantically incompatible would be considered compatible.

**Comment by ufcpp** [comment]
@jnm2 

Struct covariance proposed here is conceptually [depth subtyping of record types](https://en.wikipedia.org/wiki/Subtyping#Width_and_depth_subtyping), while interface covariance is [behavioral subtyping](https://en.wikipedia.org/wiki/Liskov_substitution_principle).

These are incompatible in terms of 'in'/'out', but, not so different in terms of subtyping.

**Comment by Neo-Ciber94** [comment]
Compare ```byref``` and ```null ref```.

```csharp
int value = 10;
ref int someRef = ref value;
ref? int otherRef = ref null;

if(ref someRef == ref otherRef) //Checking for same memory address
{
   //Do Something
}

if(ref otherRef == null) //Checking for null ref.
{
   //Do Something
}
```

**Comment by tannergooding** [comment]
@Neo-Ciber94, you can compare a `byref` to a `null ref` today by using `System.Runtime.CompilerServices.Unsafe`

You can either convert `someRef` to a pointer (using `Unsafe.AsPointer(ref someRef)`) or by converting `null` to a byref (using `Unsafe.AsRef<T>(null)`)

**Comment by alrz** [comment]
From the recent announcement [blog post](https://blogs.msdn.microsoft.com/dotnet/2018/11/12/building-c-8-0/):

>The need to keep the runtime stable has prevented us from implementing new language features in it for more than a decade. With the side-by-side and open-source nature of the modern runtimes, we feel that we can responsibly evolve them again, and do language design with that in mind.

DIM seems to be the only feature with runtime requirements in this release. Does that mean we only ship these features in major releases or also minor releases?

**Comment by Ultrahead** [comment]
Please consider #2013 (Proposal: allow generic constraints to be part of the signature for overloading resolution)

**Comment by markusschaber** [comment]
I'd suggest simple types which map to C types for interop / p/invoke.

We have some C code which uses `long` and `unsigned long` C types. While this is 32 bit on all 32 bit platforms we care about, it's different for 64 bit platforms - Windows uses 32 bit, while Linux uses 64 bit. (I did not check any other platforms yet.)

As we need to support both 32 and 64 bit Linux, we currently have 80% of our p/invoke signatures, structs etc. duplicated with if/else code in wrapper classes.

As far as I can see, those types need to be special cased in the runtime (like IntPtr), so I guess that might need a CLR extension...

**Reply by colejohnson66** [reply]
For future readers: `System.Runtime.InteropServices.CLong` exists for this purpose.

https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.clong?view=net-8.0

**Comment by gafter** [comment]
> DIM seems to be the only feature with runtime requirements in this release. Does that mean we only ship these features in major releases or also minor releases?

Major or minor releases of *what*?  I can easily imagine such changes would be introduced in a minor version of Visual Studio.  But major runtime changes seem more likely in major releases of that runtime.

**Comment by yuridiniz** [comment]
> Please consider #2013 (Proposal: allow generic constraints to be part of the signature for overloading resolution)

This solution would help me a lot

I will give the most current example that I stumbled upon it

```c#
public abstract class BasicBusiness<T, TKey> where T : class, new()
{
    public async virtual Task<T> Get(TKey key) { /*...*/ }
    public async virtual Task<T> Remove(TKey key) { /*...*/ }
    /* etc */
}

public abstract class BasicBusiness<T> : BasicBusiness<T, int>  where T : EntityIntKey, new()
{
}

public abstract class BasicBusiness<T> : BasicBusiness<T, short>  where T : EntityShortKey, new()
{
}

public abstract class BasicBusiness<T> : BasicBusiness<T, Guid>  where T : EntityGuidKey, new()
{
}

```

This would have avoided me major refactoring efforts

**Comment by LyndonGingerich** [comment]
> Please consider #2013 (Proposal: allow generic constraints to be part of the signature for overloading resolution)

Another use case (from #7902):

```csharp
public static T? GetFirst<T>(List<T> a) where T : struct => a.Count > 0 ? a[0] : null; // Member with the same signature is already declared
public static T? GetFirst<T>(List<T> a) where T : class => a.Count > 0 ? a[0] : null; // Member with the same signature is already declared
```

The semantic nullable type (including both `System.Nullable<T>` and the class `T?` notation), with null-coalescing and the `??` operator, is one of my favorite C# features. I would like to use it as much as possible rather than introducing inefficient wrapper `Option` types.

**Reply by HaloFour** [reply]
In terms of performance, I would expect a custom `Option<T>` struct to be as performant as `Nullable<T>` since it would have the same field layout.

**Reply by tannergooding** [reply]
`Nullable<T>` is intrinsic to the runtime and gets specialized support in many scenarios, including boxing, which is not available to user-defined types.

**Reply by HaloFour** [reply]
Sure it does, but from a performance point of view it should be the same and provides a workaround that can be used today given `Nullable<T>` doesn't permit reference types, the runtime doesn't allow overloading based on constraints and the C# language doesn't allow overloading based on return type.  I'm not arguing that those changes shouldn't be considered, only that there are ... options (pun intended) 😄

**Comment by louthy** [comment]
This list would be amazing:

1. Higher-kinded polymorphism (see https://github.com/dotnet/csharplang/issues/339 and https://github.com/dotnet/coreclr/issues/6524)
2. Union and intersection types (https://github.com/dotnet/csharplang/discussions/344, though likely not that syntax).

And, not sure if this is CLR or language, but generic parameters on operators.
